prometheus/cmx/client.go

```diff
@@ -0,0 +1,303 @@
+)
+
+// Client allows sending batches of Prometheus samples to CMX.
+type Client struct {
+	logger log.Logger
+
+	url        string
+	timeout    time.Duration
+	dimensions map[string]TagValue
+}
+
+var metricDefinitions map[string]bool = make(map[string]bool)
+var metricDefinitionMutex sync.RWMutex
+
+// NewClient creates a new Client.
+func NewClient(logger log.Logger, url string, timeout time.Duration, dimensions map[string]TagValue) *Client {
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	return &Client{
+		logger:     logger,
+		url:        url,
+		timeout:    timeout,
+		dimensions: dimensions,
+	}
+}
+
+type metricDefinition struct {
+	Version     int    `json:"version"`
+	MetricID    string `json:"metric-id"`
+	DisplayName string `json:"display-name"`
+	Description string `json:"description"`
+	Units       string `json:"units"`
+}
+
+// StoreMetricDefinitionsRequest is used for building a JSON request for storing samples
+// via the CMX.
+type StoreMetricDefinitionsRequest struct {
+	MetricDefinitions []metricDefinition `json:"metric-definitions"`
+}
+
+type metricSample struct {
+	MetricID   string              `json:"metric-id"`
+	Source     string              `json:"source"`
+	Timestamp  []int64             `json:"timestamp"`
+	Value      []float64           `json:"value"`
+	Dimensions map[string]TagValue `json:"dimensions"`
+	Tags       map[string]TagValue `json:"tags"`
+}
+
+// StoreSamplesRequest is used for building a JSON request for storing samples
+// via the CMX.
+type StoreSamplesRequest struct {
+	MetricSamples []metricSample `json:"metric-samples"`
+}
+
+// tagsFromMetric translates Prometheus metric into CMX tags.
+func tagsFromMetric(m model.Metric) map[string]TagValue {
+	tags := make(map[string]TagValue, len(m)-1)
+	for l, v := range m {
+		if l == model.MetricNameLabel {
+			continue
+		}
+		tags[string(l)] = TagValue(v)
+	}
+	return tags
+}
+
+func (c *Client) putMetricDef(mdreq StoreMetricDefinitionsRequest, u *url.URL) error {
+	// post metric definitions
+	buf, err := json.Marshal(mdreq)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+	level.Info(c.logger).Log("msg", "pushing new metric definitions to", "url", u.String()+"/metric", "len", len(mdreq.MetricDefinitions))
+	req, err := http.NewRequest("POST", u.String()+"/metric", bytes.NewBuffer(buf))
+	if err != nil {
+		level.Info(c.logger).Log("msg", "error post new metric definitions", "err", err)
+		return err
+	}
+	req.Header.Set("Content-Type", contentTypeJSON)
+	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
+	if err != nil {
+		level.Info(c.logger).Log("msg", "error do new metric definitions", "err", err)
+		return err
+	}
+	defer func() {
+		io.Copy(ioutil.Discard, resp.Body)
+		resp.Body.Close()
+	}()
+
+	// API returns status code 200 for successful writes.
+	// http://opentsdb.net/docs/build/html/api_http/put.html
```

Comment:
we can take out the references to opentsdb here

Reply:
addressed